File: src/gpx.ts
import type { TrackPoint } from "./types";

/**
 * Liest den ersten direkten Kind-Text eines Elements mit einem bestimmten
 * (unqualifizierten) Tag-Namen aus. GPX-Dateien nutzen üblicherweise einen
 * Default-Namespace ohne Präfix, sodass `tagName` dem lokalen Namen entspricht.
 */
function findChildText(parent: Element, tagName: string): string | null {
  for (const child of Array.from(parent.children)) {
    if (child.tagName === tagName) {
      return child.textContent;
    }
  }
  return null;
}

function parseTimeToMs(raw: string | null): number | undefined {
  if (!raw) {
    return undefined;
  }
  const ms = Date.parse(raw.trim());
  return Number.isNaN(ms) ? undefined : ms;
}

function parseEleM(raw: string | null): number | undefined {
  if (!raw) {
    return undefined;
  }
  const value = Number.parseFloat(raw.trim());
  return Number.isFinite(value) ? value : undefined;
}

function parsePoint(el: Element): TrackPoint {
  const latRaw = el.getAttribute("lat");
  const lonRaw = el.getAttribute("lon");
  const lat = latRaw !== null ? Number.parseFloat(latRaw) : NaN;
  const lon = lonRaw !== null ? Number.parseFloat(lonRaw) : NaN;

  if (!Number.isFinite(lat) || !Number.isFinite(lon)) {
    throw new Error("Ungültige Koordinaten in der GPX-Datei.");
  }

  const point: TrackPoint = { lat, lon };

  const ele = parseEleM(findChildText(el, "ele"));
  if (ele !== undefined) {
    point.ele = ele;
  }

  const time = parseTimeToMs(findChildText(el, "time"));
  if (time !== undefined) {
    point.time = time;
  }

  return point;
}

function findName(doc: Document): string | null {
  const nameEls = Array.from(doc.getElementsByTagName("name"));

  const trkName = nameEls.find((el) => el.parentElement?.tagName === "trk");
  if (trkName?.textContent?.trim()) {
    return trkName.textContent.trim();
  }

  const metaName = nameEls.find((el) => el.parentElement?.tagName === "metadata");
  if (metaName?.textContent?.trim()) {
    return metaName.textContent.trim();
  }

  return null;
}

/**
 * Parst eine GPX-1.0/1.1-Datei (String) und liefert Name sowie alle
 * Trackpunkte in Reihenfolge. Fällt auf Routenpunkte (`rtept`) zurück,
 * falls keine Trackpunkte vorhanden sind.
 */
export function parseGpx(xml: string): { name: string | null; points: TrackPoint[] } {
  const parser = new DOMParser();
  const doc = parser.parseFromString(xml, "text/xml");

  if (doc.getElementsByTagName("parsererror").length > 0) {
    throw new Error("Die GPX-Datei enthält ungültiges XML.");
  }

  const rootTag = doc.documentElement?.tagName;
  if (!rootTag || rootTag !== "gpx") {
    throw new Error("Die Datei ist keine gültige GPX-Datei.");
  }

  let pointEls = Array.from(doc.getElementsByTagName("trkpt"));
  if (pointEls.length === 0) {
    pointEls = Array.from(doc.getElementsByTagName("rtept"));
  }

  if (pointEls.length === 0) {
    throw new Error("Die GPX-Datei enthält keine Trackpunkte.");
  }

  const points = pointEls.map(parsePoint);
  const name = findName(doc);

  return { name, points };
}

function escapeXml(value: string): string {
  return value
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&apos;");
}

function buildTrkpt(point: TrackPoint): string {
  const attrs = `lat="${point.lat}" lon="${point.lon}"`;
  const children: string[] = [];

  if (point.ele !== undefined) {
    children.push(`      <ele>${point.ele}</ele>`);
  }
  if (point.time !== undefined) {
    children.push(`      <time>${new Date(point.time).toISOString()}</time>`);
  }

  if (children.length === 0) {
    return `    <trkpt ${attrs}/>`;
  }

  return `    <trkpt ${attrs}>\n${children.join("\n")}\n    </trkpt>`;
}

/**
 * Erzeugt eine valide GPX-1.1-Datei mit einem einzelnen Track/Segment.
 */
export function buildGpx(name: string, points: TrackPoint[]): string {
  const trkpts = points.map(buildTrkpt).join("\n");

  return `<?xml version="1.0" encoding="UTF-8"?>
<gpx version="1.1" creator="Trailscape" xmlns="http://www.topografix.com/GPX/1/1">
  <trk>
    <name>${escapeXml(name)}</name>
    <trkseg>
${trkpts}
    </trkseg>
  </trk>
</gpx>
`;
}

File: lib/gpx.dart
/// GPX-Import/-Export für Trailscape.
///
/// Parst GPX 1.0/1.1-Dateien (Track- und Routenpunkte) und erzeugt valide
/// GPX-1.1-Dateien aus einer Liste von [TrackPoint]s.
library;

import 'package:xml/xml.dart';

import 'models.dart';

/// Namespace der Garmin-TrackPointExtension, in der die Herzfrequenz je
/// Trackpunkt transportiert wird (`gpxtpx:hr`) — de-facto-Standard, den auch
/// Komoot, Strava & Co. beim GPX-Export verwenden.
const String _garminTrackPointExtensionNs =
    'http://www.garmin.com/xmlschemas/TrackPointExtension/v1';

/// Liest den unqualifizierten (lokalen) Tag-Namen eines Elements.
String _localName(XmlElement el) => el.name.local;

/// Sucht den ersten direkten Kind-Text eines Elements mit gegebenem
/// (unqualifiziertem) Tag-Namen.
String? _findChildText(XmlElement parent, String tagName) {
  for (final child in parent.childElements) {
    if (_localName(child) == tagName) {
      return child.innerText;
    }
  }
  return null;
}

/// Sucht den Text des ersten Nachfahren-Elements mit gegebenem
/// (unqualifiziertem) Tag-Namen — anders als [_findChildText] auch beliebig
/// tief verschachtelt, wie es `gpxtpx:hr` innerhalb von
/// `<extensions><gpxtpx:TrackPointExtension>` ist.
String? _findDescendantText(XmlElement parent, String tagName) {
  for (final el in parent.descendantElements) {
    if (_localName(el) == tagName) {
      final text = el.innerText.trim();
      if (text.isNotEmpty) return text;
    }
  }
  return null;
}

int? _parseTimeToMs(String? raw) {
  if (raw == null) return null;
  final trimmed = raw.trim();
  if (trimmed.isEmpty) return null;
  try {
    return DateTime.parse(trimmed).millisecondsSinceEpoch;
  } on FormatException {
    return null;
  }
}

double? _parseEleM(String? raw) {
  if (raw == null) return null;
  final trimmed = raw.trim();
  if (trimmed.isEmpty) return null;
  final value = double.tryParse(trimmed);
  if (value == null || !value.isFinite) return null;
  return value;
}

int? _parseHrBpm(String? raw) {
  if (raw == null) return null;
  final trimmed = raw.trim();
  if (trimmed.isEmpty) return null;
  final value = double.tryParse(trimmed);
  if (value == null || !value.isFinite) return null;
  return value.round();
}

TrackPoint _parsePoint(XmlElement el) {
  final latRaw = el.getAttribute('lat');
  final lonRaw = el.getAttribute('lon');
  final lat = latRaw != null ? double.tryParse(latRaw.trim()) : null;
  final lon = lonRaw != null ? double.tryParse(lonRaw.trim()) : null;

  if (lat == null || !lat.isFinite || lon == null || !lon.isFinite) {
    throw const FormatException('Ungültige Koordinaten in der GPX-Datei.');
  }

  return TrackPoint(
    lat: lat,
    lon: lon,
    ele: _parseEleM(_findChildText(el, 'ele')),
    time: _parseTimeToMs(_findChildText(el, 'time')),
    hr: _parseHrBpm(_findDescendantText(el, 'hr')),
  );
}

String? _findName(XmlDocument doc) {
  final nameEls = doc.findAllElements('name');

  for (final el in nameEls) {
    final parent = el.parent;
    if (parent is XmlElement && _localName(parent) == 'trk') {
      final text = el.innerText.trim();
      if (text.isNotEmpty) return text;
    }
  }

  for (final el in nameEls) {
    final parent = el.parent;
    if (parent is XmlElement && _localName(parent) == 'metadata') {
      final text = el.innerText.trim();
      if (text.isNotEmpty) return text;
    }
  }

  return null;
}

/// Parst eine GPX-1.0/1.1-Datei (String) und liefert Name sowie alle
/// Trackpunkte in Reihenfolge. Fällt auf Routenpunkte (`rtept`) zurück,
/// falls keine Trackpunkte vorhanden sind.
({String? name, List<TrackPoint> points}) parseGpx(String xmlString) {
  final XmlDocument doc;
  try {
    doc = XmlDocument.parse(xmlString);
  } on XmlException {
    throw const FormatException('Die GPX-Datei enthält ungültiges XML.');
  }

  final root = doc.rootElement;
  if (_localName(root) != 'gpx') {
    throw const FormatException('Die Datei ist keine gültige GPX-Datei.');
  }

  var pointEls = doc.findAllElements('trkpt').toList();
  if (pointEls.isEmpty) {
    pointEls = doc.findAllElements('rtept').toList();
  }

  if (pointEls.isEmpty) {
    throw const FormatException('Die GPX-Datei enthält keine Trackpunkte.');
  }

  final points = pointEls.map(_parsePoint).toList();
  final name = _findName(doc);

  return (name: name, points: points);
}

/// Erzeugt eine valide GPX-1.1-Datei mit einem einzelnen Track/Segment.
///
/// Trägt ein Trackpunkt eine Herzfrequenz ([TrackPoint.hr]), wird sie als
/// Garmin-TrackPointExtension (`gpxtpx:hr`) mitgeschrieben — das Format, das
/// auch Komoot, Strava und die meisten Sportuhren beim GPX-Export nutzen und
/// das beim erneuten Einlesen (siehe [parseGpx]) wieder erkannt wird.
///
/// [time] wird — falls angegeben — zusätzlich als `<metadata><time>`
/// geschrieben (z. B. der Aufnahmezeitpunkt einer Tour).
String buildGpx(String name, List<TrackPoint> points, {DateTime? time}) {
  final hasHr = points.any((p) => p.hr != null);

  final builder = XmlBuilder();
  builder.processing('xml', 'version="1.0" encoding="UTF-8"');
  builder.element(
    'gpx',
    attributes: {
      'version': '1.1',
      'creator': 'Trailscape',
      'xmlns': 'http://www.topografix.com/GPX/1/1',
      if (hasHr) 'xmlns:gpxtpx': _garminTrackPointExtensionNs,
    },
    nest: () {
      builder.element(
        'metadata',
        nest: () {
          builder.element('name', nest: name);
          if (time != null) {
            builder.element(
              'time',
              nest: time.toUtc().toIso8601String(),
            );
          }
        },
      );
      builder.element(
        'trk',
        nest: () {
          builder.element('name', nest: name);
          builder.element(
            'trkseg',
            nest: () {
              for (final point in points) {
                builder.element(
                  'trkpt',
                  attributes: {
                    'lat': '${point.lat}',
                    'lon': '${point.lon}',
                  },
                  nest: () {
                    if (point.ele != null) {
                      builder.element('ele', nest: '${point.ele}');
                    }
                    if (point.time != null) {
                      final iso = DateTime.fromMillisecondsSinceEpoch(
                        point.time!,
                        isUtc: true,
                      ).toIso8601String();
                      builder.element('time', nest: iso);
                    }
                    if (point.hr != null) {
                      builder.element(
                        'extensions',
                        nest: () {
                          builder.element(
                            'gpxtpx:TrackPointExtension',
                            nest: () {
                              builder.element(
                                'gpxtpx:hr',
                                nest: '${point.hr}',
                              );
                            },
                          );
                        },
                      );
                    }
                  },
                );
              }
            },
          );
        },
      );
    },
  );

  final document = builder.buildDocument();
  return '${document.toXmlString(pretty: true, indent: '  ')}\n';
}
